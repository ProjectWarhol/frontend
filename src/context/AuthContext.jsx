import * as SecureStore from 'expo-secure-store'
import createDataContext from "./createDataContext"
import Api from '../api/api'
import { navigate } from '../navigationRef'

const authReducer = (state, action) => {
  switch (action.type){
    case 'add_error':
      return {...state, errorMessage: action.payload }
    case 'login':
      return {errorMessage: '', cookie: action.payload}
    case 'clear_error_message':
      return { ...state, errorMessage: '' }
    case 'logout':
      return { cookie: null, errorMessage: '' }
    default:
      return state
  }
}

const logout = dispatch => async () => {
  try { const cookie = await SecureStore.getItemAsync('cookie')
  const config = {headers:{
    'my.sid': cookie
  }}
  await Api.post('/users/logout', config)
  await SecureStore.deleteItemAsync('cookie')
  dispatch({ type: 'logout'})
  navigate('unAuthenticatedUser')
}catch(err){
  const cookie = await SecureStore.getItemAsync('cookie')
    await SecureStore.deleteItemAsync('cookie')
    dispatch({ type: 'logout'})
    navigate('unAuthenticatedUser')
}}

const tryLocalLogin = dispatch => async () => {
  const cookie = await SecureStore.getItemAsync('cookie')
if(cookie && isCookieValid()){
  dispatch({ type: 'login', payload: cookie})
  navigate('mainFlow')
}else{
  navigate('unAuthenticatedUser')
}
}

const isCookieValid = async(cookie) => {
  const config = {headers:{
    'my.sid': cookie
  }}
  await Api.get('/users/session', config)
  return true
}

const clearErrorMessage = dispatch => () => {
  dispatch({ type: 'clear_error_message' })
}

const login = (dispatch) => async ({ email, password }) => {
    try{
      const response = await Api.post('/users/login', { userCredential: email, password: password })
      let cookieArray = response["headers"]["set-cookie"]
      let cookieString = cookieArray.toString()
      const cookie = cookieString.substring(
        cookieString.indexOf('=') + 1,
        cookieString.indexOf(';')
      )
      await SecureStore.setItemAsync('cookie', cookie)
      dispatch({ type: 'login', payload: cookie})
      navigate('mainFlow')
    } catch (err){
      {dispatch({ type: 'add_error',  payload: 'Something went wrong' }, console.log(err))}
    }
  }

export const { Provider, Context } = createDataContext(
  authReducer,
  { login, logout, clearErrorMessage, tryLocalLogin },
  { cookie: null, errorMessage: '' }
)
