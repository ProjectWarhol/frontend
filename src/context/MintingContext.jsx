import createDataContext from "./createDataContext"
import Api from '../api/api'
import { navigate } from '../navigationRef'

const mintingReducer = (state, action) => {
  switch (action.type){
    case 'image':
      return {...state, image: action.payload}
    default:
      return state
  }
}

const takePicture = dispatch => async (camera) => {
  if(camera){
    const data = await camera.takePictureAsync(null)
    dispatch({type: 'image', payload: data.uri} )
    navigate('image')
  }
}

export const { Provider, Context } = createDataContext(
  mintingReducer,
  { takePicture },
  { image: null }
)
