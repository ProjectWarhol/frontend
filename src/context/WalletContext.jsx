import createDataContext from "./createDataContext"
import Api from '../api/api'
import { navigate } from '../navigationRef'
// import {createCustodialWallet, storeCustodialWallet, generateSeedPhrase} from '../wallet/CustodialWallet'

const walletReducer = (state, action) => {
  switch (action.type){
    case('createWallet'):
      return {...state }
    case 'signup':
      return {...state }
    case 'clear_error_message':
      return { ...state, errorMessage: '' }
    case 'add_error':
      return {...state, errorMessage: action.payload }
    default:
      return state
  }
}

const signup = (dispatch) => {
  return async ({ email, password, repeatedPassword, userName }) => {
    try{
      if (password === repeatedPassword)
      {const response = await Api.post('/users/createUser', { email: email, password: password, userName: userName })
      console.log(response.data)
      dispatch({ type: 'signup' })
      navigate('selectWallet')}
      else{dispatch({type: 'add_error', payload: 'Passwords need to match'})}
    } catch (err)
    {dispatch({ type: 'add_error', payload: 'Something went wrong' }, console.log(err))
    }
  }
}

const clearErrorMessage = dispatch => () => {
  dispatch({ type: 'clear_error_message' })
}

export const { Provider, Context } = createDataContext(
  walletReducer,
  { signup, clearErrorMessage },
  { errorMessage: '', email: '', password: '', userName: ''}
)
