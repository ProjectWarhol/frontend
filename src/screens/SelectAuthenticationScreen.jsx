import React, { useContext } from 'react'
import {  StyleSheet, View, SafeAreaView } from 'react-native'
import { Text, Button } from 'react-native-elements'
import Spacer from '../components/Spacer'
import { AuthContext } from '../context/AuthContext'
import { NavigationEvents } from 'react-navigation'

const SelecteAuthenticationScreen = ({navigation}) => {
  const { clearErrorMessage } = useContext(AuthContext)

  return (
  <SafeAreaView style={styles.container}>
  <NavigationEvents
  onWillBlur={clearErrorMessage}
  />
  <View style={styles.headerContainer}>
  <Text style={styles.text} h3>Welcome</Text>
  </View>
  <Spacer/>
  <View style={styles.buttonParent}>
  <View style={styles.loginButtonContainer}>
  <Button
  testID='login'
  title='Log in'
  style={styles.button}
  onPress={()=>{navigation.navigate('login')}}
  />
  <Text style={styles.buttonDescription}>I have an Account</Text>
  </View >
  <View style={styles.signupButtonContainer}>
  <Button
  testID='signup'
  title='Sign up'
  style={styles.button}
  onPress={()=>{navigation.navigate('signup')}}
  />
  <Text style={styles.buttonDescription}>I want to create an Account</Text>
  </View>
  </View>
  </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container:{
    display: 'flex',
  },
  headerContainer:{
    justifyContent:'center',
    flexDirection: 'row',
    marginTop: '50%'
  },
  text: {
    fontSize: 30,
  },
  buttonParent:{
    flexDirection: 'row',
  },
  loginButtonContainer:{
    flexDirection: 'column',
    flex:1,
    marginLeft: 20,
    marginRight: 10
  },
  signupButtonContainer:{
    flexDirection: 'column',
    flex:1,
    marginLeft: 10,
    marginRight: 20
  },
  buttonDescription:{
    marginLeft: 20,
    marginRight: 20,
    paddingTop: 10
  }
})

export default SelecteAuthenticationScreen
