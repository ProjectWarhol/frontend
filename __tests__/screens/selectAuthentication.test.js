import React from "react";
import { create } from "react-test-renderer";
import SelectAuthenticationScreen from "../../src/screens/SelectAuthenticationScreen";
import { AuthProvider } from '../../src/context/AuthContext';

const navigation = {
  navigate: jest.fn()
}

jest.mock('react-navigation', () => ({
  withNavigation: SelectAuthenticationScreen => props => (
    <SelectAuthenticationScreen navigation={{ navigate: jest.fn() }} {...props} />
  ), NavigationEvents: 'mockNavigationEvents'
}));

const tree = create(<AuthProvider value={null}><SelectAuthenticationScreen navigation={navigation}/></AuthProvider>)

test('snapshot', ()=>{
  expect(tree).toMatchSnapshot()
})

test('press login button', ()=>{
  const loginButton = tree.root.findByProps({ testID: 'login'}).props
  loginButton.onPress()
  expect(navigation.navigate).toBeCalledWith('login')
})

test('press signup button', ()=>{
  const signupButton = tree.root.findByProps({ testID: 'signup'}).props
  signupButton.onPress()
  expect(navigation.navigate).toBeCalledWith('signup')
})
