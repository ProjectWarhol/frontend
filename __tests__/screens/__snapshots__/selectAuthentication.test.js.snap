// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <mockNavigationEvents
    onWillBlur={[Function]}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": "50%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#242424",
          "fontSize": 36,
        }
      }
    >
      Welcome
    </Text>
  </View>
  <View
    style={
      Object {
        "margin": 15,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
          "marginLeft": 20,
          "marginRight": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="login"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2089dc",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Log in
            </Text>
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#242424",
            "marginLeft": 20,
            "marginRight": 20,
            "paddingTop": 10,
          }
        }
      >
        I have an Account
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
          "marginLeft": 10,
          "marginRight": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="signup"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2089dc",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Sign up
            </Text>
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#242424",
            "marginLeft": 20,
            "marginRight": 20,
            "paddingTop": 10,
          }
        }
      >
        I want to create an Account
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
